host/class/uac/usb_host_uac/include/usb/uac.h

```diff
@@ -0,0 +1,480 @@
+    UAC_OUTPUT_TERMINAL_COMMUNICATION_SPEAKER         = 0x0306,
+    UAC_OUTPUT_TERMINAL_LOW_FREQUENCY_EFFECTS_SPEAKER = 0x0307
+} uac_output_terminal_type_t;
+
+/******************************** Refer frmts10.pdf ***************************************************/
+
+/**
+ * @brief Audio Data Format Type I Codes
+ *
+ * @see Table A 1.1 of frmts10.pdf
+ */
+typedef enum {
+    UAC_TYPE_I_UNDEFINED                              = 0x0000,
+    UAC_TYPE_I_PCM                                    = 0x0001,
+    UAC_TYPE_I_PCM8                                   = 0x0002,
+    UAC_TYPE_I_IEEE_FLOAT                             = 0x0003,
+    UAC_TYPE_I_ALAW                                   = 0x0004,
+    UAC_TYPE_I_MULAW                                  = 0x0005
+} uac_type_i_format_t;
+
+/**
+ * @brief Format Type Codes
+ *
+ */
+typedef enum {
+    UAC_FORMAT_TYPE_UNDEFINED                         = 0x0000,
+    UAC_FORMAT_TYPE_I                                 = 0x0001,
+    UAC_FORMAT_TYPE_II                                = 0x0002,
+    UAC_FORMAT_TYPE_III                               = 0x0003
+} uac_format_type_t;
+
+/**
+ * @brief Audio Class-Specific Interface Descriptor
+ *
+ */
+typedef struct {
+    uint8_t bLength;
+    uint8_t bDescriptorType;
+    uint8_t bDescriptorSubtype;
+} __attribute__((packed)) uac_desc_header_t;
+
+/**
+ * @brief Audio Class-Specific AC Interface Header Descriptor (bInCollection=2)
+ *
+ * @see Table 4-2 of audio10.pdf
+ */
+typedef struct {
+    uint8_t bLength;
+    uint8_t bDescriptorType;
+    uint8_t bDescriptorSubtype;
+    uint16_t bcdADC;
+    uint16_t wTotalLength;
+    uint8_t bInCollection;
+    uint8_t baInterfaceNr[2];
+} __attribute__((packed)) uac_ac_header_desc_t;
+
+/**
+ * @brief Audio Class-Specific AC Input Terminal Descriptor
+ *
+ * @see Table 4-3 of audio10.pdf
+ */
+typedef struct {
+    uint8_t bLength;
+    uint8_t bDescriptorType;
+    uint8_t bDescriptorSubtype;
+    uint8_t bTerminalID;
+    uint16_t wTerminalType;
+    uint8_t bAssocTerminal;
+    uint8_t bNrChannels;
+    uint16_t wChannelConfig;
+    uint8_t iChannelNames;
+    uint8_t iTerminal;
+} __attribute__((packed)) uac_ac_input_terminal_desc_t;
+
+/**
+ * @brief Audio Class-Specific AC Output Terminal Descriptor
+ *
+ * @see Table 4-4 of audio10.pdf
+ */
+typedef struct {
+    uint8_t bLength;
+    uint8_t bDescriptorType;
+    uint8_t bDescriptorSubtype;
+    uint8_t bTerminalID;
+    uint16_t wTerminalType;
+    uint8_t bAssocTerminal;
+    uint8_t bSourceID;
+    uint8_t iTerminal;
+} __attribute__((packed)) uac_ac_output_terminal_desc_t;
+
+/**
+ * @brief Audio Class-Specific AC Feature Unit Descriptor (ch=2, bControlSize=2)
+ *
+ * @see Table 4-7 of audio10.pdf
+ */
+typedef struct {
+    uint8_t bLength;
+    uint8_t bDescriptorType;
+    uint8_t bDescriptorSubtype;
+    uint8_t bUnitID;
+    uint8_t bSourceID;
+    uint8_t bControlSize;
+    uint8_t bmaControls[3 * 2];
```

Comment:
It's quite tricky descriptor, I can't define it well in a struct. `bmaControls` size is variable, it depends on the audio `bNrChannels` and `bControlSize`, here we assume (for most device) the audio device is dual-channel (3=2+1 (chanel 0)), each channel's `bControlSize` is 2 bytes. 

But it's should no problem for any other case. Because the real descriptor is parsed as it is